Mathlib/Analysis/Calculus/SecondDerivativeTest.lean

```diff
@@ -0,0 +1,141 @@
+
+
+section SecondDeriv
+
+variable {f : ℝ → ℝ} {x₀ : ℝ}
+
+/-- If the slope from a critical point `x₀` to `b > x₀` is positive then so is the derivative
+ at `b`. -/
+lemma deriv_pos_of_lt_of_slope_pos {b : ℝ} (hb : x₀ < b) (hbf : 0 < slope (deriv f) x₀ b)
+    (hf : deriv f x₀ = 0) : 0 < deriv f b := by
+  unfold slope at hbf
+  rw [hf] at hbf
+  simp_all
+
+/-- If the slope from a critical point `x₀` to `b < x₀` is positive then the derivative
+ at `b` is negative. -/
+lemma deriv_neg_of_gt_of_slope_pos {b : ℝ} (hb : b < x₀) (hbf : 0 < slope (deriv f) x₀ b)
+    (hf : deriv f x₀ = 0) : deriv f b < 0 := by
+  unfold slope at hbf
+  simp_rw [smul_eq_mul, hf] at hbf
+  rw [mul_comm] at hbf
+  have : 0 < deriv f b / (b - x₀) := sub_zero (deriv f b) ▸ hbf
+  contrapose this
+  exact not_lt.mpr <| div_nonpos_of_nonneg_of_nonpos (not_lt.mp this) (by linarith)
+
+/-- If the derivative is nonzero in a (specific) punctured neighborhood then the
+function is differentiable in a punctured neighborhood. -/
+theorem eventually_differentiable_of_deriv_nonzero {ε : ℝ}
+    (hε : ε > 0)
+    (hε₀ : ∀ b ∈ Ioo (x₀ - ε) x₀, deriv f b ≠ 0)
+    (hε₁ : ∀ b ∈ Ioo x₀ (x₀ + ε), deriv f b ≠ 0) :
+    ∀ᶠ x in 𝓝[≠] x₀, DifferentiableAt ℝ f x :=
+    (eventually_mem_set.mpr <| insert_mem_nhds_iff.mp <| insert_Ioo₀ hε hε ▸
+    Ioo_mem_nhds (by linarith) (by linarith)).mono
+    fun _ hb => differentiableAt_of_deriv_ne_zero <| hb.elim (hε₀ _) (hε₁ _)
+
+
+/-- If `f''(x) > 0` then `f' < 0` on an interval to the left of `x`. -/
+lemma deriv_neg_of_deriv_deriv_pos (hf : deriv (deriv f) x₀ > 0)
+    (hd : deriv f x₀ = 0) : ∃ u < x₀, ∀ b ∈ Ioo u x₀, deriv f b < 0 := by
+  obtain ⟨u,hu⟩ := (mem_nhdsWithin_Iio_iff_exists_mem_Ico_Ioo_subset
+    (show x₀ - 1 < x₀ by simp)).mp
+      <| nhds_left'_le_nhds_ne x₀ <| (tendsto_nhds.mp <| hasDerivAt_iff_tendsto_slope.mp
+      (differentiableAt_of_deriv_ne_zero <| ne_of_gt hf).hasDerivAt) (Ioi 0) isOpen_Ioi hf
+  exact ⟨u, hu.1.2, fun b hb => deriv_neg_of_gt_of_slope_pos hb.2 (hu.2 hb) hd⟩
+
+
+/-- If `f''(x) > 0` then `f' > 0` on an interval to the right of `x`. -/
+lemma deriv_pos_of_deriv_deriv_pos (hf : deriv (deriv f) x₀ > 0)
+    (hd : deriv f x₀ = 0) : ∃ u > x₀, ∀ b ∈ Ioo x₀ u, deriv f b > 0 := by
+  obtain ⟨u,hu⟩ := (mem_nhdsWithin_Ioi_iff_exists_mem_Ioc_Ioo_subset (show x₀ < x₀ + 1 by simp)).mp
+    <| nhds_right'_le_nhds_ne x₀ <|(tendsto_nhds.mp <| hasDerivAt_iff_tendsto_slope.mp
+    (differentiableAt_of_deriv_ne_zero <| ne_of_gt hf).hasDerivAt) (Ioi 0) isOpen_Ioi hf
+  exact ⟨u, hu.1.1, fun b hb => deriv_pos_of_lt_of_slope_pos hb.1 (hu.2 hb) hd⟩
+
+/-- If `f''(x) > 0` then `f'` changes sign at `x`.
+This lemma applies to functions like `x^2 + 1[x ≥ 0]` as well as twice differentiable
+functions.
+-/
+lemma deriv_neg_pos_of_deriv_deriv_pos
+    (hf : deriv (deriv f) x₀ > 0) (hd : deriv f x₀ = 0) :
+    ∃ ε > 0, (∀ b ∈ Ioo (x₀-ε) x₀, deriv f b < 0) ∧
+              ∀ b ∈ Ioo x₀ (x₀ + ε), 0 < deriv f b := by
```

Comment:
I think this should be phrased as "eventually in a punctured neighborhood, the sign of `deriv f b` matches the sign of `b - x`."